### PR TITLE
MINOR FIX: Casting Issue

### LINKED ISSUE #6: MINOR FIX: Casting Issue
When UpsertDataList is called after AddData there is a cast issue.  This code rub is to address that bug.

```
            var xeption = new Xeption();

            xeption.AddData(
                key: keyName,
                values: textRequiredValue);

            xeption.UpsertDataList(
                key: keyName,
                value: textLengthValue);
```

image

### PR DESCRIPTION
**Pull request by cjdutoit**
closes #6

### CODE REVIEW

File: Xeption.Tests/XeptionTests.Logic.cs [hunk 1]
```diff
@@ -85,6 +85,40 @@ public void ShouldAppendListOfKeyValues()
             }
         }
 
+        [Fact]
+        public void ShouldAppendListOfKeyValuesWhenUpsertDataListIsCalledAfterAddData()
+        {
+            string keyName = "Name";
+            string textRequiredValue = "Text is required";
+            string textLengthValue = "Text must be between 2 and 25 characters";
```
Comment (hassanhabib) [inline Xeption.Tests/XeptionTests.Logic.cs:93]: Why 2 to 25?

**Comment by cjdutoit** [comment]
@hassanhabib not sure if this PR is stuck.  Not sure why there is pending on my comments?

**Comment by hassanhabib** [comment]
@cjdutoit I think is this because you selected a full review so you add a bunch of comments then later add an entire review.

**Comment by cjdutoit** [comment]
@hassanhabib ok, I thought I had to do that to start the workflow again on a change. more familiar with DevOps, but getting used to GitHub now.  Are you able to merge/approve this?

File: Xeption/Xeption.cs [hunk 2]
```diff
@@ -38,7 +39,17 @@ public void UpsertDataList(string key, string value)
         {
             if (this.Data.Contains(key))
             {
-                (this.Data[key] as List<string>)?.Add(value);
+                if(this.Data[key] is Array)
```
Comment (hassanhabib) [inline Xeption/Xeption.cs:42]: @cjdutoit what's a good scenario for this?
You are checking here if the value for the key is an array. then appending to that array. Where would we need to use this? Isn't this what `AddData` does anyway with `params`?

**Comment by cjdutoit** [comment]
@hassanhabib - In the unit tests we use `.AddData(...)` and I know the correct and cleaner way to do this would be like this:

image

but when I started out with the new validation standard I sort of followed the flow of adding the structural validation items at the top and then working my way down the logical validation like this

image

I was expecting the `UpsertDataList ` to just work here based on its name, but then ran into the casting issue as the value could not be converted to a list with `(this.Data[key] as List<string>)?.Add(value);` as it was initialised as a `string[]` in `.AddData(...)` resulting in `null` value as in the screenshot below (and no item being added to the data dictionary)

image

This fix is to just fool proof it as I expect others will also hit this same scenario as I have, so I do think there is a use case for this as people would expect to be able to perform an UpsertDataList in any state that the data dictionary may have.

The check to see if it is an array `string[]` and if it is we then cast it to `List<string>`  where we can append items to.

I had to do the cast like this `(this.Data[key] as string[]).ToList();`  
as `(this.Data[key] as List<string>)?.Add(value);` results in null value if the data type is `string[]`
(not sure if there is a better way of doing this)

**Comment by hassanhabib** [comment]
> @hassanhabib - In the unit tests we use `.AddData(...)` and I know the correct and cleaner way to do this would be like this:
> 
> image
> 
> but when I started out with the new validation standard I sort of followed the flow of adding the structural validation items at the top and then working my way down the logical validation like this
> 
> image
> 
> I was expecting the `UpsertDataList ` to just work here based on its name, but then ran into the casting issue as the value could not be converted to a list with `(this.Data[key] as List<string>)?.Add(value);` as it was initialised as a `string[]` in `.AddData(...)` resulting in `null` value as in the screenshot below (and no item being added to the data dictionary)
> 
> image
> 
> This fix is to just fool proof it as I expect others will also hit this same scenario as I have, so I do think there is a use case for this as people would expect to be able to perform an UpsertDataList in any state that the data dictionary may have.
> 
> I had to do the cast like this `(this.Data[key] as string[]).ToList();` as `(this.Data[key] as List<string>)?.Add(value);` results in null value if the data type is `string[]` (not sure if there is a better way of doing this)

@cjdutoit Let's find sometime to discuss this. hit me up on LinkedIn.

**Comment by cjdutoit** [comment]
@hassanhabib I had another look at the code and found a much simpler fix to this issue.   By changing the parameter array to a list before adding we now avoid the casting issue.

**Comment by cjdutoit** [comment]
@hassanhabib anything else needed here or is this ok now?

**Pull request closed without merging**